beacon-chain/endtoend/evaluators.go

```diff
@@ -0,0 +1,144 @@
+package endtoend
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	ptypes "github.com/gogo/protobuf/types"
+	"github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// policy defines an function that returns a bool
+// for whether the evaluator it is assigned to runs.
+type policy func(currentEpoch uint64) bool
+
+// evaluation defines a function that takes in BeaconChainClient
+// and uses it to monitor the chain, to determine where it fulfills the conditions required.
+type evaluation func(client eth.BeaconChainClient) error
+
+// evaluator defines a struct for executing a given evaluation
+// on a running E2E test given that the policy returns true.
+type evaluator struct {
+	name       string
+	policy     policy
+	evaluation evaluation
+}
+
+// runEvaluators takes in testing.T, BeaconChainClient and an array of evaluators.
+// Using the client to check the state, it runs each evaluator in a subtest.
+func runEvaluators(t *testing.T, client eth.BeaconChainClient, evaluators []evaluator) {
+	in := new(ptypes.Empty)
+	chainHead, err := client.GetChainHead(context.Background(), in)
+	if err != nil {
+		t.Errorf("failed to get chain head: %v", err)
+	}
+	currentEpoch := chainHead.BlockSlot / params.BeaconConfig().SlotsPerEpoch
+
+	for _, evaluator := range evaluators {
+		if evaluator.policy(currentEpoch) {
+			fmt.Printf("Running %s\n", evaluator.name)
+			t.Run(evaluator.name, func(t *testing.T) {
+				if err := evaluator.evaluation(client); err != nil {
+					t.Fatal(err)
+				}
+			})
+		}
+	}
+}
+
+func afterNEpochs(epochs uint64) func(uint64) bool {
+	return func(currentEpoch uint64) bool {
+		return currentEpoch == epochs
```

Comment:
Shouldn't this be >= ?

Reply:
It's `==` so the evaluator doesn't run for every epoch after the one it was meant for.

Reply:
That doesn't make sense though. If the current epoch is greater than `epochs`, wouldn't you want to return true?

Reply:
The way the testing is designed, it only queries the evaluators once per epoch. So by using == it is only runs once for that epoch.

Although changing the design to run tests for every epoch isn't a bad idea, to get more coverage.